src/electron/process_manager.ts

```diff
@@ -272,100 +176,130 @@ export class ConnectionManager {
+  async start(checkConnectivity: boolean) {
+    if (checkConnectivity) {
+      try {
+        this.isUdpEnabled = await this.checkConnectivity();
+      } catch (e) {
+        this.resolveStop();
+        throw e;
       }
-    };
+    }
 
-    // We have to listen for both events: error means the process could not be launched and in that
-    // case exit will not be invoked.
-    this.process.on('error', onExit.bind((this)));
-    this.process.on('exit', onExit.bind((this)));
+    if (this.process.isRunning) {
+      // Restart once the current process exits without notifying the exit.
+      this.stop(false);
+      await this.process.onExit;
+      console.log('restarting tun2socks');
+    }
+
+    this.process.launch(this.getProcessArgs());
+
+    return new Promise((resolve, reject) => {
+      // Declare success when tun2socks is running.
+      this.process.onStderr = (data?: string | Buffer) => {
+        if (data && data.toString().includes('tun2socks running')) {
+          this.process.onStderr = undefined;
+          this.notifyStop = true;
+          if (isWindows) {
+            powerMonitor.once('suspend', this.suspendListener);
+            powerMonitor.once('resume', this.resumeListener);
+          }
+          resolve();
+        }
+      };
+
+      // Reject on early exit. On routine exit, the promise will be already resolved.
+      this.process.onExit.then((code?: number) => {
+        console.log('tun2socks exited with code', code);
+        if (!this.notifyStop) {
+          return;
+        }
+        this.resolveStop();
+        reject(errors.fromErrorCode(code || errors.ErrorCode.UNEXPECTED));
+      });
+    });
   }
 
-  // Use #onExit to be notified when the process exits.
-  stop() {
-    if (!this.process) {
-      // Never started.
-      if (this.exitListener) {
-        this.exitListener();
-      }
-      return;
+  // Stops the tun2socks process. Notifies the caller when the process has exited
+  // by resolving `onceStopped` if `notify` is true.
+  stop(notify=true) {
+    if (isWindows) {
+      powerMonitor.removeListener('suspend', this.suspendListener);
+      powerMonitor.removeListener('resume', this.resumeListener);
     }
-
-    this.process.kill();
+    this.notifyStop = notify;
+    this.process.stop();
   }
 
-  set onExit(newListener: (() => void)|undefined) {
-    this.exitListener = newListener;
+  get onceStopped(): Promise<void> {
+    return this.stopped;
   }
-}
 
-class SsLocal extends ChildProcessHelper {
-  constructor(private readonly proxyPort: number) {
-    super(pathToEmbeddedBinary('shadowsocks-libev', 'ss-local'));
+  // Notifes tun2socks that network connectivity changed. Checks whehter UDP support changed,
+  // restarting if it did.
+  async networkChanged() {
+    if (this.isUdpEnabled === await this.checkConnectivity()) {
+      return;
+    }
+    this.isUdpEnabled = !this.isUdpEnabled;
```

Comment:
This is clever, but I think it would be better to store `await this.checkConnectivity()` in a temp variable.

As an example of how this could go wrong, consider starting with `isUdpEnabled = true`.  Two `networkChanged` events arrive in rapid succession, triggering two calls to `checkConnectivity`.  If the second call resolves first, with value `false`, this code will flip `isUdpEnabled` to `false`.  When the first call resolves later, also with value `false`, this code will flip `isUdpEnabled` again, to `true`!

Reply:
Done. We now update `isUdpEnabled` in `checkConnectivity`, caching the previous result as per your suggestion.